src/controller/lintController.ts

```diff
@@ -6,15 +6,83 @@ import { CredentialController } from "./credentialController";
 import { DirectoryController } from "./directoryController";
 import constants from "../config/config";
 import { Message, MessageType, ErrorMessages } from "../types";
+import { LintView } from "../views/lintView";
 import { NotificationView } from "../views/notificationView";
 
 export class LintController {
+  private dirtyFiles: Set<string>;
+
   constructor(
     private diagnosticCollection: vscode.DiagnosticCollection,
     private credentialController: CredentialController,
     private addonCacheController: AddonCacheController,
     private directoryController: DirectoryController
-  ) {}
+  ) {
+    this.dirtyFiles = new Set();
+  }
+
+  /**
+   * Tracks currently dirtied files.
+   * @param event A vscode event describing a document change.
+   */
+  toggleDirty(event: vscode.TextDocumentChangeEvent) {
+    const document = event.document;
+    if (document.isDirty) {
+      this.dirtyFiles.add(document.uri.fsPath);
+      // if the change was an empty undo, it was reset to its natural state
+    } else if (event.reason && event.contentChanges) {
+      this.dirtyFiles.delete(document.uri.fsPath);
+    }
+  }
+
+  /**
+   * Remove a dirtied file.
+   */
+  removeDirty(document: vscode.TextDocument) {
+    this.dirtyFiles.delete(document.uri.fsPath);
+  }
+
+  /**
+   * Clears existing lints if a document is saved when it was dirtied.
+   */
+  clearLintsOnDirty(document: vscode.TextDocument) {
+    if (this.dirtyFiles.has(document.uri.fsPath)) {
+      this.clearLints(document.uri);
+    }
+    this.removeDirty(document);
+  }
+
+  /**
+   * If a file is deleted from a version, remove the version's lints.
+   * @param event A VS Code FileDeleteEvent
+   */
+  // Since lints only happen if the workspace is open directly to the version folder,
+  // the diagnosticCollection can be cleared without fear of edge cases.
```

Comment:
Nit: any reason to mix comment styles?

Reply:
it's something I picked up from courses, where information that isn't immediately relevant to the documentation but still notable are left out. Since this looks like a bug from an outsider perspective, I added that.